pkg/controller/replication_controller_utils.go

```diff
@@ -0,0 +1,270 @@
+func (r *RCExpectations) SetExpectations(rc *api.ReplicationController, add, del int) error {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return err
+	}
+	return r.Add(&PodExpectations{add: int64(add), del: int64(del), key: rcKey})
+}
+
+// Decrements the expectation counts of the given rc.
+func (r *RCExpectations) lowerExpectations(rc *api.ReplicationController, add, del int) {
+	if podExp, exists, err := r.GetExpectations(rc); err == nil && exists {
+		podExp.Seen(int64(add), int64(del))
+	}
+}
+
+// CreationObserved atomically decrements the `add` expecation count of the given replication controller.
+func (r *RCExpectations) CreationObserved(rc *api.ReplicationController) {
+	r.lowerExpectations(rc, 1, 0)
+}
+
+// DeletionObserved atomically decrements the `del` expectation count of the given replication controller.
+func (r *RCExpectations) DeletionObserved(rc *api.ReplicationController) {
+	r.lowerExpectations(rc, 0, 1)
+}
+
+// Expectations are either fulfilled, or expire naturally.
+type Expectations interface {
+	Fulfilled() bool
+}
+
+// PodExpectations track pod creates/deletes.
+type PodExpectations struct {
+	add int64
+	del int64
+	key string
+}
+
+// Seen decrements the add and del counters.
+func (e *PodExpectations) Seen(add, del int64) {
+	atomic.AddInt64(&e.add, -add)
+	atomic.AddInt64(&e.del, -del)
+}
+
+// Fulfilled returns true if this expectation has been fulfilled.
+func (e *PodExpectations) Fulfilled() bool {
+	// TODO: think about why this line being atomic doesn't matter
+	return atomic.LoadInt64(&e.add) <= 0 && atomic.LoadInt64(&e.del) <= 0
+}
+
+// NewRCExpectations returns a store for PodExpectations.
+func NewRCExpectations() *RCExpectations {
+	return &RCExpectations{cache.NewTTLStore(expKeyFunc, ExpectationsTimeout)}
+}
+
+// PodControlInterface is an interface that knows how to add or delete pods
+// created as an interface to allow testing.
+type PodControlInterface interface {
+	// createReplica creates new replicated pods according to the spec.
+	createReplica(namespace string, controller api.ReplicationController) error
+	// deletePod deletes the pod identified by podID.
+	deletePod(namespace string, podID string) error
+}
+
+// RealPodControl is the default implementation of PodControllerInterface.
+type RealPodControl struct {
+	kubeClient client.Interface
+	recorder   record.EventRecorder
+}
+
+func (r RealPodControl) createReplica(namespace string, controller api.ReplicationController) error {
+	desiredLabels := make(labels.Set)
+	for k, v := range controller.Spec.Template.Labels {
+		desiredLabels[k] = v
+	}
+	desiredAnnotations := make(labels.Set)
+	for k, v := range controller.Spec.Template.Annotations {
+		desiredAnnotations[k] = v
+	}
+
+	createdByRef, err := api.GetReference(&controller)
+	if err != nil {
+		return fmt.Errorf("unable to get controller reference: %v", err)
+	}
+
+	createdByRefJson, err := json.Marshal(createdByRef)
```

Comment:
Unversioned serialization alert - if you want to fix this in a follow up that's fine (since you didn't introduce the problem with this change) but please add a big TODO.